odoo/tools/sql.py

```diff
@@ -363,3 +365,53 @@ def increment_fields_skiplock(records, *fields):
     cr = records._cr
     cr.execute(query, {'ids': records.ids})
     return bool(cr.rowcount)
+
+
+def value_to_translated_trigram_pattern(value):
+    """ Escape value to match a translated field's trigram index content
+
+    The trigram index function jsonb_path_query_array("column_name", '$.*')::text
+    concatenates all translations' representations as the indexed text. So the
+    original text needs to be JSON-escaped correctly to match it.
+
+    :param str value: value provided in domain
+    :return: a pattern to match the indexed text
+    """
+    if len(value) < 3:
+        # matching less than 3 characters will not take advantage of the index
+        return '%'
+
+    # apply JSON escaping to value
+    json_escaped = json.dumps(value)[1:-1]
+
+    # apply PG wildcard escaping to JSON-escaped text
+    wildcard_escaped = re.sub(r'(_|%|\\)', r'\\\1', json_escaped)
+
+    # add wildcards around it to get the pattern
+    return f"%{wildcard_escaped}%"
+
+
+def pattern_to_translated_trigram_pattern(pattern):
+    """ Escape pattern to match a translated field's trigram index content
+
+    The trigram index function jsonb_path_query_array("column_name", '$.*')::text
+    concatenates all translations' representations as the indexed text. So the
+    original pattern needs to be JSON-escaped correctly to match it.
+
+    :param str pattern: value provided in domain
+    :return: a pattern to match the indexed text
+    """
+    # find the parts around (non-escaped) wildcard characters (_, %)
+    sub_patterns = re.findall(r'((?:.|\n)*?(?<!\\)(?:\\\\)*)(?:_|%|$)', pattern)
```

Comment:
Comment would probably be clearer if this said something along the lines of "split on wildcard characters". Also maybe use `re.VERBOSE` and add comments to the regex? And could use `re.DOTALL` and remove the `\n` from the leading form.

Other possible alternative: process the entire thing by hand? Might be able to fold the next step in as well. And it would avoid a polynomial evaluation vulnerability.